src/Storages/Hive/HiveCommon.cpp

```diff
@@ -0,0 +1,243 @@
+#include <ThriftHiveMetastore.h>
+#include <thrift/protocol/TBinaryProtocol.h>
+#include <thrift/transport/TBufferTransports.h>
+#include <thrift/transport/TSocket.h>
+#include <Storages/Hive/HiveCommon.h>
+
+#if USE_HIVE
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int NO_HIVEMETASTORE;
+    extern const int BAD_ARGUMENTS;
+}
+
+bool HiveMetastoreClient::shouldUpdateTableMetadata(
+    const String & db_name, const String & table_name, const std::vector<Apache::Hadoop::Hive::Partition> & partitions)
+{
+    String cache_key = getCacheKey(db_name, table_name);
+    std::shared_ptr<HiveMetastoreClient::HiveTableMetadata> metadata = table_metadata_cache.get(cache_key);
+    if (!metadata)
+        return true;
+
+    auto old_partiton_infos = metadata->getPartitionInfos();
+    if (old_partiton_infos.size() != partitions.size())
+        return true;
+
+    for (const auto & partition : partitions)
+    {
+        auto it = old_partiton_infos.find(partition.sd.location);
+        if (it == old_partiton_infos.end())
+            return true;
+
+        const auto & old_partition_info = it->second;
+        if (!old_partition_info.haveSameParameters(partition))
+            return true;
+    }
+    return false;
+}
+
+std::shared_ptr<HiveMetastoreClient::HiveTableMetadata>
+HiveMetastoreClient::getTableMetadata(const String & db_name, const String & table_name)
+{
+    LOG_TRACE(log, "Get table metadata for {}.{}", db_name, table_name);
+    std::lock_guard lock{mutex};
+
+    auto table = std::make_shared<Apache::Hadoop::Hive::Table>();
+    std::vector<Apache::Hadoop::Hive::Partition> partitions;
+    try
+    {
+        client->get_table(*table, db_name, table_name);
+
+        /// Query the latest partition info to check new change.
+        client->get_partitions(partitions, db_name, table_name, -1);
+    }
+    catch (apache::thrift::transport::TTransportException & e)
+    {
+        setExpired();
+        throw Exception(ErrorCodes::NO_HIVEMETASTORE, "Hive Metastore expired because {}", String(e.what()));
+    }
+
+    bool update_cache = shouldUpdateTableMetadata(db_name, table_name, partitions);
+    String cache_key = db_name + "." + table_name;
+
+    std::shared_ptr<HiveMetastoreClient::HiveTableMetadata> metadata = table_metadata_cache.get(cache_key);
```

Comment:
Same as comment a little above:
```suggestion
    HiveTableMetadataPtr metadata = table_metadata_cache.get(cache_key);
```
(Please also change it everywhere)

Reply:
And probably it should be like 
```suggestion
    std::shared_ptr<HiveMetastoreClient::HiveTableMetadata> metadata;
```
and assigned to `table_metadata_cache.get(cache_key)` in `else` below.

Reply:
I am not quite understand about this. I think whether we need to update cache or not, we should try to get metadata cache.
- If update_cache is false, we need to create new `HiveTableMetadata` object, but we can still reuse some `PartitionInfo` objects from returned cache entry.  
- Otherwise return the entry we getted from cache directly.